### PR TITLE
switch from publicsuffix to psl

### PR DESCRIPTION
**Pull request by rushmorem**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -48,7 +49,7 @@ http = "0.2"
 httparse = "1"
 httpdate = "0.3.2"
 once_cell = "1"
-publicsuffix = { default-features = false, version = "1.5.4" }
+publicsuffix = "2.0.11"
```
Comment (rushmorem) [inline Cargo.toml:52]: @algesten `publicsuffix` v2 is very light on dependencies but we can trim them even further by disabling default features if you only need to lookup unicode domains. Alternatively, we can use the `psl` crate which doesn't have any dependencies (except for `psl-types` which is already being pulled in by `addr`) but that doesn't allow the list to be updated at runtime as it's static. Finally we can ship with no PSL implementation at all which allows users to use any implementation of `psl_types::List` they like but that's a breaking change.

**Review by algesten — APPROVED** [review]
Thanks for doing this!

My only reason for using publicsuffix was to avoid saving cookies for TLD. Since I want hreq to "just work", I'm interested in any solution that:

1. Have a minimum of dependencies.
2. Hardcodes a "reasonable" TLD list during compilation.

Being able to supply an override list instead of the compiled in one, would be nice, but an optional extra.

**Comment by rushmorem** [comment]
> Thanks for doing this!

My pleasure :)

> My only reason for using publicsuffix was to avoid saving cookies for TLD

That was my first impression too but [this test](https://github.com/algesten/hreq/blob/a62f256a1c4d57595d4dd6afc69f1fc40c751d57/src/client/cookies.rs#L214) failed when not checking whether the domain is valid first so I thought you wanted domain validation too. That's why I included `addr`. Would you prefer I remove `_tcp.example.com` from those tests or change it to `true`?

> Have a minimum of dependencies

Since you don't need domain validation in `is_valid_cookie_domain` and you prefer a hardcoded list, `psl` alone will be sufficient for this. It only depends on `psl-types` which is very small and has no dependencies. It actually used to be part of `psl` but I pulled out those types so they can be shared with `publicsuffix` to make switching between the two seamless.

> Hardcodes a "reasonable" TLD list during compilation

`psl` compiles the entire public suffix list down to native Rust code. This is done when publishing the crate, not during compilation, so compile times are still fast. As [per crates.io](https://crates.io/crates/psl), the crate is 178 kB. Is that OK with you?

> Being able to supply an override list instead of the compiled in one, would be nice, but an optional extra.

That would require using `publicsuffix`. Since you prefer fewer dependencies, I suggest we go with `psl` which won't give the ability to supply a different list. We can always switch to `publicsuffix` later if that becomes a requirement.

**Comment by algesten** [comment]
@rushmorem `_tcp.example.com` I guess is a weird one, and I'm not particularly married to the idea of having an implicit domain validation as part of the cookie handling. `psl` does sound ideal, I think we should switch!

**Comment by rushmorem** [comment]
> `psl` does sound ideal, I think we should switch!

@algesten Done.

**Review by algesten — APPROVED** [review]
Looks good! Let's merge!

**Comment by algesten** [comment]
And a big thanks!

**Comment by rushmorem** [comment]
Don't mention it :)

**Pull request merged**